### PR TITLE
build: refactor the typescript compiler broccoli plugin

### PR DESCRIPTION
**Pull request by hansl**

### CODE REVIEW

**Comment by zackarychapple** [comment]
This is failing lint.  Other than that LGTM.

File: lib/broccoli/broccoli-typescript.js [hunk 1]
```diff
@@ -177,52 +131,56 @@ class DiffingTSCompiler extends Plugin {
...
+
+      // Add file registry.
+    var allFiles = program.getSourceFiles();
+
+    // Find the entry, update the registry.
+    let allEntries = this.listEntries().filter(entry => {
+      return true;
+      // return allFiles.some(source => {
+      //   return path.resolve(this.inputPaths[0], source.fileName) == path.resolve(this.inputPaths[0], entry.relativePath);
+      // });
```
Comment (gkalpak) [inline lib/broccoli/broccoli-typescript.js]: Is this commented block left here on purpose ?

**Comment by jkuri** [comment]
Very good job @hansl, what concerns the code I have no comments. The incremental build as well as the initial build works flawlessly. The incremental build takes from 200-400ms which is really an improvement and makes development apps fun.

There are some bugs that needs to be resolved;

- Fix the `eslint` errors
```
/Users/jan/Desktop/angular-cli/lib/broccoli/broccoli-typescript.js
141:48  error  'entry' is defined but never used  no-unused-vars
```
- The `.ts` files are copied into `dist` folder instead of compiled (`.js`) files. When fixing this make sure that `.ts` are not copied next to `.js` ones, they are not needed in the `dist` directory.

When this will be resolved I think its ready to merge.

**Comment by hansl** [comment]
The `.ts` files being copied is a separate issue (part of angular2-app). I'll do in a separate PR.

**Comment by jkuri** [comment]
Great job! LGTM.

**Pull request merged**